feat: Add support for webhook listeners

come-nc opened the pull request:
## Summary

Add support for registering webhooks as listeners of events.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

come-nc committed:
feat: Add support for webhook listeners

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
feat: Serialize event data and author userid to webhook

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

ChristophWurst reviewed (commented):
Code looks good! How will this be used? Can we assume that apps know their webhook listeners when the app is just *registered* and the system isn't fully booted yet?

come-nc commented:
> Code looks good! How will this be used?

Plan is to use that for integration of an external workflow engine, but architecture is not final yet.
One problem is we’d prefer if the webhook gets called only on certains conditions, but then we end up with `apps/workflowengine` complexity.

> Can we assume that apps know their webhook listeners when the app is just _registered_ and the system isn't fully booted yet?

Hum, what is not accessible at that point?
If I want to have an app which search in its database and user settings to know which webhooks to register, will that work?

ChristophWurst commented:
> > Can we assume that apps know their webhook listeners when the app is just _registered_ and the system isn't fully booted yet?
> 
> Hum, what is not accessible at that point? If I want to have an app which search in its database and user settings to know which webhooks to register, will that work?

That works. Anything that is provided by other apps might be missing. Say information from a user backend.

If you always want to wire a specific event to  webhook listeners it's fine. But the lack of this information could be critical when you need to conditionally register webhooks. I haven't yet understood the full picture of the new automations so this might not be necessary.

https://docs.nextcloud.com/server/latest/developer_manual/app_development/bootstrap.html#nextcloud-20-and-later for my original ideas and thoughts about the two step register/boot process.

come-nc committed:
feat: Add webhooks application

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
Webhooks are moved to an application with registrations in DB instead. They will also need to support filtering of events triggerring the webhook.

- [ ] Skeleton app
- [ ] API to register webhooks
- [ ] occ command to list them
- [ ] Actually call the webhooks, through background jobs
- [ ] Serialize the event to the webhook
- [ ] Add a filtering system

come-nc commented:
- [ ] Store app id from appapi header if available

come-nc committed:
Revert "feat: Serialize event data and author userid to webhook"

This reverts commit 45d3bd6e30e112a89ffe2fcfdd3ba4bbdc384346.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Revert "feat: Add support for webhook listeners"

This reverts commit 3b790df0127b2bf95e8fe3a8460aa3813e58bef8.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
feat: Make node events serializable

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
feat: Add command to list registered webhooks

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
feat: Add admin delegation for webhooks

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
feat: Add event filtering to webhooks

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix: Add event filter to openapi

Somehow this does not work in the ocs_api_viewer app, so we should look
 into better specifying format for eventFilter.
Ideally it should also be set as in body, not query.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
chore: Add tests for the PHPMongoQuery class

It appears that it does not match Mongo current documentation exactly so
 we should look into adapting it. Having equality autodetect regex is a
 bit weird.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
feat: Add app_api app id to saved information about webhook

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
feat: Add support for headers and authentication headers in webhooks

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix: Fix TODOs about authentication data, list possible auth methods

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

marcelklehr commented on apps/webhooks/lib/ResponseDefinitions.php:
```diff
@@ -0,0 +1,26 @@
+ *     httpMethod: string,
+ *     uri: string,
+ *     event?: string,
+ *     eventFilter?: array<mixed>,
```
Is this still failing?

come-nc replied:
Changed it to `array<string,mixed>`. It’s not exactly right because you can have an array of filters with integer keys I think, but it’s close enough and psalm type system does not allow us to define nested arrays correctly.

come-nc commented:
Follow-up:

- [ ] Extract user filter from eventFilter so that the listener is not even registered depending on logged in user
- [ ] Same for group membership

come-nc committed:
fix: Fix eventFilter definition in ResponseDefinitions

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

marcelklehr commented on apps/webhooks/lib/ResponseDefinitions.php:
```diff
@@ -16,7 +16,7 @@
  *     uri: string,
  *     event?: string,
- *     eventFilter?: array<mixed>,
+ *     eventFilter?: array<string,mixed>,
```
It might be possible to define a recursive type, to avoid the "mixed" here, but I'm not sure

come-nc replied:
https://github.com/vimeo/psalm/issues/5739 nope

julien-nc commented on apps/webhooks/lib/Controller/WebhooksController.php:
```diff
@@ -0,0 +1,218 @@
+		?array $eventFilter,
+		?array $headers,
+		?string $authMethod,
+		?array $authData,
```
Are we sure that authData will always be an array? Even with other future auth methods?

julien-nc commented on apps/webhooks/lib/Db/WebhookListener.php:
```diff
@@ -0,0 +1,96 @@
+		return AuthMethod::from(parent::getAuthMethod());
+	}
+
+	public function getAuthDataClear(): array {
+		if ($this->authData === null) {
+			return [];
+		}
+		return json_decode($this->crypto->decrypt($this->getAuthData()), associative:true, flags:JSON_THROW_ON_ERROR);
+	}
+
+	public function setAuthDataClear(?array $data): void {
```
Nitpicking: I think it's enough to call those methods `getAuthData` and `setAuthData`. I feel like current names are ambiguous. Or maybe `getDecryptedAuthData` and `setEncryptedAuthData`. Not a big deal anyway.

come-nc replied:
It’s not possible, `getAuthData` and `setAuthData` are the magical setter/getter for the DB column. (And they are used by the mapper so we cannot make them encrypt/decrypt).
If we want to use those names I need to rename the DB column to `auth_data_encrypted` or something. Do you want me to do that? I’m fine with both approach.

julien-nc replied:
Oh right, no let's not rename the DB columns. Maybe we can find more explicit names. Maybe current names are fine... :grin:

julien-nc commented on apps/webhooks/lib/Db/WebhookListenerMapper.php:
```diff
@@ -0,0 +1,159 @@
+		return $this->findEntities($qb);
+	}
+
+	public function addWebhookListener(
```
Missing some method return types and PhpDoc blocks (useful for `@throws` IMO).

julien-nc commented on apps/webhooks/lib/Listener/WebhooksEventListener.php:
```diff
@@ -0,0 +1,90 @@
+		if ($event instanceof \JsonSerializable) {
+			return $event;
+		} else {
+			/* Event is not serializable, we fallback to reflection to still send something */
```
Noice!

julien-nc commented on apps/webhooks/lib/Migration/Version1000Date20240527153425.php:
```diff
@@ -0,0 +1,72 @@
+
+		if (!$schema->hasTable(WebhookListenerMapper::TABLE_NAME)) {
+			$table = $schema->createTable(WebhookListenerMapper::TABLE_NAME);
+			$table->addColumn('id', Types::INTEGER, [
```
```suggestion
			$table->addColumn('id', Types::BIGINT, [
```

julien-nc commented on apps/webhooks/lib/Service/PHPMongoQuery.php:
```diff
@@ -0,0 +1,340 @@
+	 * @throws Exception
+	 */
+	public static function find(array $query, array $documents, array $options = []): array {
+		if(empty($documents) || empty($query)) {
```
php-cs will probably complain about the syntax. Should we avoid checking this file or fix the code style?

come-nc replied:
https://github.com/nextcloud/server/actions/runs/9367363495/job/25786877013?pr=45475
`php-cs` is happy, I ran `cs:fix` on the file already.

julien-nc reviewed (changes requested):
Works great! A few minor comments and questions.

nickvergessen commented on apps/webhooks/lib/AppInfo/Application.php:
```diff
@@ -0,0 +1,55 @@
+		foreach ($configuredEvents as $eventName) {
+			// $logger->error($eventName.' '.\OCP\Files\Events\Node\NodeWrittenEvent::class, ['exception' => new \Exception('coucou')]);
+			$dispatcher->addServiceListener(
+				$eventName,
```
Should we block certain event names and allow apps to defined blocked events?
E.g. login events should not send the users password to other servers.

julien-nc replied:
Nice catch, yes, all authentication-related events should be "blacklisted". It's difficult to list all those dispatched by community apps but those should be enough:
* all auth events from core
* user_ldap
* user_saml
* user_oidc

@nickvergessen Do you think of any other kind of event that should be avoided?

come-nc replied:
That’s a good point.
Then I think we should remove the fallback and only allow `\JsonSerializable` implementors to be listened to.
Then an application can decide to not implement it to disallow webhooks or to only serialize some properties (ie not the password for PostLoginEvent). @julien-nc @nickvergessen what do you think?

nickvergessen replied:
> or to only serialize some properties (ie not the password for PostLoginEvent).

Sounds good.

nickvergessen commented on apps/webhooks/lib/Migration/Version1000Date20240527153425.php:
```diff
@@ -0,0 +1,72 @@
+			]);
+			$table->addColumn('uri', Types::STRING, [
+				'notnull' => true,
+				'length' => 256,
```
In talk we used 4k length for the url. Not sure 256 might be enough in all cases

come-nc commented on apps/webhooks/appinfo/info.xml:
```diff
@@ -0,0 +1,33 @@
+	</dependencies>
+
+	<commands>
+		<command>OCA\Webhooks\Command\Index</command>
```
`ListCommand`, like `user:list`?

julien-nc replied:
Yes, `ListCommand` is more explicit, I think `ListWebhooks` matches more with `webhooks:list`.

come-nc commented on apps/webhooks/lib/BackgroundJobs/WebhookCall.php:
```diff
@@ -0,0 +1,56 @@
+			}
+			$response = $client->request($webhookListener->getHttpMethod(), $webhookListener->getUri(), $options);
+			$statusCode = $response->getStatusCode();
+			if ($statusCode >= 200 && $statusCode < 300) {
```
Yeah I messed it up, inverting the branch content, nice catch.

come-nc commented on apps/webhooks/lib/Controller/WebhooksController.php:
```diff
@@ -0,0 +1,218 @@
+			throw new OCSBadRequestException($e->getMessage(), $e);
+		} catch (\DomainException $e) {
+			throw new OCSForbiddenException($e->getMessage(), $e);
+		} catch (Exception $e) {
```
Was supposed to be `\Exception` like the other methods, fixing it.

come-nc committed:
fix: Fix errors spotted by reviewers, fix `@throws` annotations

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

provokateurin commented:
I'm unsubscribing from this PR for now, please ping me if you need a review of the (Open)API.

come-nc committed:
chore: rename Index.php to ListWebhooks.php

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented on apps/webhooks/lib/AppInfo/Application.php:
```diff
@@ -0,0 +1,55 @@
+		$mapper = $container->get(WebhookListenerMapper::class);
+
+		/* Listen to all events with at least one webhook configured */
+		$configuredEvents = $mapper->getAllConfiguredEvents();
```
Yes, see https://github.com/nextcloud/server/blob/9fc8959d9d236667c4761f11d7e935c22a374f0b/apps/webhooks/lib/Db/WebhookListenerMapper.php#L134
For the cache to make sense it needs to be in session, should I add that in the mapper or in the `Application` class?

nickvergessen replied:
> For the cache to make sense it needs to be in session, should I add that in the mapper or in the Application class?

I prefer to keep the mapper plain, to know when a SQL query is fired.
If you use this in more places in the future you could move it to a service file.
So for now I would put it into Application

come-nc replied:
See https://github.com/nextcloud/server/pull/45475/commits/57c5ddbd207fa8b7565f49a6e1ca2fa85814603d , but if this caching was done in the mapper instead, I could reset the cache whenever a webhook is inserted or updated, what do you think?

nickvergessen replied:
so time for a service? :P 
Or the mapper is the service.
Could also have a private uncached function there which is called by the "public" cached function.

come-nc replied:
I prefer if the mapper is the service, splitting this in two classes sounds overkill to me, + it weakens the idea that all db access should go through these functions. (what I mean is that if the caching is in the service it might miss calls directly on the mapper).

The Application class cannot use DI, so moving the cache to the mapper. Incomming.

come-nc replied:
https://github.com/nextcloud/server/pull/45475/commits/0cbb36e232ddb91950f226cd2ea35ccf46b3c7c7

come-nc committed:
fix: Make webhook event serialization opt-in with a new interface

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix: Cache webhooks listened events for 5min

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix: Move caching to the mapper instead

Application class cannot use DI, and having the cache in the mapper
 allows for invalidating it when inserting or updating a webhook
 registration.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

nickvergessen commented on apps/webhooks/lib/BackgroundJobs/WebhookCall.php:
```diff
@@ -0,0 +1,56 @@
+		[$data, $webhookId] = $argument;
+		$webhookListener = $this->mapper->getById($webhookId);
+		$client = $this->clientService->newClient();
+		$options = [];
```
might make sense to add verify with our certificate manager as well:
```
'verify' => $this->certificateManager->getAbsoluteBundlePath(),
```

Just learned about this today, but it then allows to easily also set up requests against self-signed servers by using the `occ security:certificate:import` command

nickvergessen commented on apps/webhooks/lib/Db/WebhookListenerMapper.php:
```diff
@@ -0,0 +1,201 @@
+		?array $eventFilter,
+		?array $headers,
+		AuthMethod $authMethod,
+		?array $authData,
```
Should we start adding:
```suggestion
		#[\SensitiveParameter]
		?array $authData,
```
for new things at least?

marcelklehr commented on apps/webhooks/lib/Migration/Version1000Date20240527153425.php:
```diff
@@ -0,0 +1,72 @@
+			]);
+			$table->addColumn('uri', Types::STRING, [
+				'notnull' => true,
+				'length' => 4096,
```
$ php occ app:enable webhooks
Column "oc_webhook_listeners"."uri" is type String, but exceeding the 4.000 length limit.

marcelklehr commented on apps/webhooks/lib/Controller/WebhooksController.php:
```diff
@@ -0,0 +1,218 @@
+				$event,
+				$eventFilter,
+				$headers,
+				AuthMethod::from($authMethod ?? AuthMethod::None->value),
```
This causes a HTTP 500 if $authMethod is not recognized

marcelklehr commented:
It seems the node events that you added the interface to are not actually emitted at the moment. I can only see the old `\\OCP\\Files::preWrite` ones

come-nc commented:
> It seems the node events that you added the interface to are not actually emitted at the moment. I can only see the old `\\OCP\\Files::preWrite` ones

They are, I tested with a NodeWrittenEvent and the webhook was called. Please detail what scenario you tested and which event firing was missing and I’ll look into it.

come-nc commented:
- [ ] webhooks app id is already used on appstore, we should rename.

marcelklehr commented on apps/webhooks/lib/BackgroundJobs/WebhookCall.php:
```diff
@@ -0,0 +1,56 @@
+					break;
+				case AuthMethod::Header:
+					$authHeaders = $webhookListener->getAuthDataClear();
+					$options['headers'] = array_merge($options['headers'], $authHeaders);
```
"Exception": "TypeError",
    "Message": "array_merge(): Argument #1 must be of type array, null given",
    "Code": 0,
    "Trace": [
      {
        "file": "/home/marcel/code/server/apps/webhooks/lib/BackgroundJobs/WebhookCall.php",
        "line": 42,
        "function": "array_merge"
      },

come-nc committed:
fix: Remove superfluous antislash from event name to avoid problems

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(webhooks): fix crash when headers is null with header auth method

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(webhooks): Refuse webhooks registration if auth header is used without auth data

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(webhooks): Add Nextcloud certificate bundle to client options

This way importing a certificate with occ security:certificate:import
 will allow to use it for webhooks.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(webhooks): Fix migration uri length

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
feat(webhooks): Add \SensitiveValue attribute to authData parameter in all classes

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(webhooks): Fix tests

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
chore(webhooks): Rename webhooks application to webhook_listeners

There is already a webhooks application in the appstore

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(webhooks): Fix a few of psalm errors

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented on apps/webhook_listeners/lib/Controller/WebhooksController.php:
```diff
@@ -0,0 +1,220 @@
+				$event,
+				$eventFilter,
+				$headers,
+				AuthMethod::from($authMethod ?? AuthMethod::None->value),
```
I want the valueerror, I just need to catch it and turn in into the proper OCS api response.

come-nc committed:
fix(webhooks): Fix openapi reponse definition

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

marcelklehr commented:
> They are, I tested with a NodeWrittenEvent and the webhook was called. Please detail what scenario you tested and which event firing was missing and I’ll look into it.

Nevermind, as discussed, it was the leading backslash that caused this

come-nc committed:
fix(webhooks): Fix last psalm and openapi problems with the API

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(webhooks): Fix a few more psalm notices

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

nickvergessen commented on lib/public/EventDispatcher/IWebhookCompatibleEvent.php:
```diff
@@ -0,0 +1,24 @@
+ *
+ * @since 30.0.0
+ */
+interface IWebhookCompatibleEvent {
```
@come-nc Please document in https://docs.nextcloud.com/server/latest/developer_manual/app_publishing_maintenance/app_upgrade_guide/upgrade_to_30.html#id1

(all changes in lib/public/)
